File: geekbang/algorithm-models/3.6- BFS 拓扑排序模板/topsort.java
// Java
// LeetCode210 课程表II


class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        // 初始化
        n = numCourses;
        edges = new ArrayList<List<Integer>>();
        inDeg = new int[n];
        for (int i = 0; i < n; i++) {
           edges.add(new ArrayList<Integer>());
           inDeg[i] = 0;
        }
        // 建图，加边
        for (int[] pre : prerequisites) {
            int ai = pre[0];
            int bi = pre[1];
            addEdge(bi, ai);
        }
        // 拓扑排序
        return topsort();
    }


    int[] topsort() {
        int[] order = new int[n];
        int m = 0;
        Queue<Integer> q = new LinkedList<Integer>();
        // 零入度点入队
        for (int i = 0; i < n; i++)
            if (inDeg[i] == 0) q.offer(i);
        while (!q.isEmpty()) {
            Integer x = q.poll();
            order[m] = x;
            m++;
            // 扩展每个点
            for (Integer y : edges.get(x)) {
                inDeg[y]--;
                if (inDeg[y] == 0) q.offer(y);
            }
        }
        // n个课程都进出过队列，说明可以完成
        if (m == n) return order;
        return new int[0];
    }


    private void addEdge(int x, int y) {
        edges.get(x).add(y);
        inDeg[y]++;
    }


    private int n;
    private List<List<Integer>> edges;
    private int[] inDeg;
}

File: geekbang/algorithm-models/3.6- BFS 拓扑排序模板/topsort.cpp
//C/C++
// LeetCode210 课程表II


class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        n = numCourses;
        // 出边数组初始化，n个空list
        edges = vector<vector<int>>(n, vector<int>());
        inDeg = vector<int>(n, 0);
        for (vector<int>& pre : prerequisites) {
            int ai = pre[0];
            int bi = pre[1];
            // 加边模板
            addEdge(bi, ai);
        }
        auto ans = topsort();
        if (ans.size() < n) return {}; // 不能完成所有课程
        return ans;
    }


private:
    // 有向图找环 模板（拓扑排序）
    // 返回学的课程数
    vector<int> topsort() {
        vector<int> order;
        // 拓扑排序基于BFS，需要队列
        queue<int> q;
        // 从所有零入度点出发
        for (int i = 0; i < n; i++)
            if (inDeg[i] == 0) q.push(i);
        // 执行BFS
        while (!q.empty()) {
            int x = q.front(); // 取队头（这门课学了）
            q.pop();
            order.push_back(x);
            // 考虑x的所有出边
            for (int y : edges[x]) {
                inDeg[y]--; // 去掉约束关系
                if (inDeg[y] == 0) {
                    q.push(y);
                }
            }
        }
        return order;
    }


    void addEdge(int x, int y) {
        edges[x].push_back(y);
        inDeg[y]++;
    }


    int n;
    vector<vector<int>> edges;
    vector<int> inDeg; // in degree 入度
};
